pkg/ruler/compat.go

```diff
@@ -57,14 +59,16 @@ type RulesLimits interface {
 
 // engineQueryFunc returns a new query function using the rules.EngineQueryFunc function
 // and passing an altered timestamp.
-func engineQueryFunc(engine *logql.Engine, overrides RulesLimits, checker readyChecker, userID string) rules.QueryFunc {
+func engineQueryFunc(cfg Config, engine *logql.Engine, overrides RulesLimits, checker readyChecker, userID string, logger log.Logger, rng *rand.Rand) rules.QueryFunc {
 	return rules.QueryFunc(func(ctx context.Context, qs string, t time.Time) (promql.Vector, error) {
 		// check if storage instance is ready; if not, fail the rule evaluation;
 		// we do this to prevent an attempt to append new samples before the WAL appender is ready
 		if !checker.isReady(userID) {
 			return nil, errNotReady
 		}
 
+		applyJitter(rng, cfg.EvaluationJitter, log.With(logger, "query_hash", logql.HashQueryString(qs)))
```

Comment:
Is this called once for a query out each time it's executed? I'm asking because we'd just need some delay at the start.

Reply:
I covered this in "Special notes for your reviewer"

Reply:
It would, but only slightly; it's a tradeoff because executing all the rules concurrently would lead to the same thing but with a lot more contention.

https://github.com/grafana/loki/pull/8129 addresses this, and will not really require jitter unless at large scales.